File: trunk/staff/tools/codegen/template/jsclient/EnumSerialization.js
#ifeq($(Enum.Extern),0) // do not serialize extern type

// enum $(Enum.NsName.!mangle)
var o$(Enum.NsName.!mangle) = staff.object("$(Enum.NsName.!dot)");
#var EnumValue 0
#foreach $(Enum.Members)
o$(Enum.NsName.!mangle).$(Member.Name) =\
#ifneq($(Member.Value),)
 $(Member.Value);
#var EnumValue $(Member.Value.!inc)
#else
 $($EnumValue);
#var EnumValue $($EnumValue.!inc)
#ifeqend
#end

o$(Enum.NsName.!mangle).serialize = function(oElement, oType)
{
#ifeq($(Enum.Options.*baseType),string)
  var sResult = "<INVALID>";
  switch (oType)
  {
#foreach $(Enum.Members)
  case $(Enum.NsName.!mangle).$(Member.Name):
    sResult = "$(Member.Name)";
    break;
#end
  };
  oElement.setText(sResult)
#else
  oElement.setText(oType)
#ifeqend
  return oElement;
}

o$(Enum.NsName.!mangle).deserialize = function(oElement)
{
#ifeq($(Enum.Options.*baseType),string)
  var oResult = -1;
  switch (staff.getElementValue(oElement))
  {
#foreach $(Enum.Members)
  case "$(Member.Name)":
    oResult = $(Enum.NsName.!mangle).$(Member.Name);
    break;
#end
  };
  return oResult;
#else
  return parseInt(staff.getElementValue(oElement));
#ifeqend
}

#ifeqend //ifeq((Enum.Extern),0) // do not serialize extern type

File: staff/tools/codegen/template/common/EnumSerialization.cpp
#cginclude "StringTypes"

std::string& operator<<(std::string& sResult, const $(Enum.NsName) eEnumValue)
{
#ifeq($(Enum.Options.*baseType),$($sStringTypes))
  switch (eEnumValue)
  {
#foreach $(Enum.Members)
  case $(Enum.Namespace)$(Member.Name):
    sResult = "$(Member.Value||Member.Name)";
    break;
#end
  default:
    RISE_THROWS(rise::CLogicNoItemException,
       "Value out of range while serializing enum [$(Enum.NsName)]: " + rise::ToStr(eEnumValue));
  };
#else
  sResult = staff::ToString(static_cast<int>(eEnumValue));
#ifeqend
  return sResult;
}

DataObject& operator<<(DataObject& rdoParam, const $(Enum.NsName) eEnumValue)
{
  std::string sParam;
  rdoParam.SetText(sParam << eEnumValue);
  return rdoParam;
}

const std::string& operator>>(const std::string& sParam, $(Enum.NsName)& reEnumValue)
{
#ifeq($(Enum.Options.*baseType),$($sStringTypes))
#foreach $(Enum.Members)
  if (sParam == "$(Member.Value||Member.Name)")
  {
    reEnumValue = $(Enum.Namespace)$(Member.Name);
  }
  else
#end
  {
    RISE_THROWS(rise::CLogicNoItemException,
       "Value out of range while deserializing enum [$(Enum.NsName)]: [" + sParam + "]");
  }
#else
  int nValue = 0;
  rdoParam.GetValue(nValue);
  reEnumValue = static_cast< $(Enum.NsName) >(nValue);
#ifeqend
}

const DataObject& operator>>(const DataObject& rdoParam, $(Enum.NsName)& reEnumValue)
{
  rdoParam.GetText() >> reEnumValue;
  return rdoParam;
}
